pytorch_pfn_extras/profiler/_tracing.py

```diff
@@ -0,0 +1,230 @@
+
+
+class ChromeTracer(Tracer):
+    """Tracer object that outputs a timeline in Chrome format.
+
+    Args:
+        max_event_count (int): Limit the amount of events that can be traced,
+            optional.
+        enable (bool): Sets the tracer in active state. Optional,
+            defaults to ``True``.
+    """
+
+    def __init__(
+        self,
+        max_event_count: Optional[int] = None,
+        enable: bool = True,
+    ) -> None:
+        self._enable = enable
+        self._event_list: List[Dict[str, Union[str, int, float]]] = []
+        self._max_event_count = max_event_count or float("inf")
+        self._event_count = 0
+        # Detect if i am a forked process, in such case I send the event to
+        # The parent process
+        self._is_cuda_available = torch.cuda.is_available()
+        self._tracer_queue: _util._QueueWorker = _util._QueueWorker(
+            self.add_remote_event, 1000
+        )
+        self._tracer_queue.initialize()
+
+    @contextlib.contextmanager
+    def add_event(self, name: str) -> Generator[None, None, None]:
+        if (
+            not self._enable
+            or not _enabled
+            or not getattr(_thread_local, "enable", True)
+        ):
+            yield
+            return
+
+        pid = os.getpid()
+        is_forked = pid != _main_pid
+        is_cuda_available = self._is_cuda_available and not is_forked
+        begin_ns = time.perf_counter_ns()
+        try:
+            yield
+        finally:
+            if self._enable and self._event_count < self._max_event_count:
+                if is_cuda_available:
+                    torch.cuda.synchronize()  # Wait for process to complete
+                self._event_count += 1
+                duration_ns = time.perf_counter_ns() - begin_ns
+                tid = threading.get_native_id()
+                # Append is thread safe so this should be fine to execute
+                # without a lock, the trace does not require the events
+                # to be ordered
+                event = dict(
+                    name=name,
+                    cat="",
+                    ph="X",
+                    ts=begin_ns / 1000,  # nano sec -> micro sec
+                    dur=duration_ns / 1000,  # ditto
+                    pid=pid,
+                    tid=tid,
+                )
+                if is_forked:
+                    self._tracer_queue.put(name, event)
+                else:
+                    self._event_list.append(
+                        cast(Dict[str, Union[str, int, float]], event)
+                    )
+
+    def add_remote_event(
+        self, name: str, event: Dict[str, Union[str, int, float]]
+    ) -> None:
+        self._event_list.append(event)
+
+    def flush(self, filename: str, writer: Writer) -> None:
+        if not self._enable:
+            return
+        # TODO(ecastill): try to work on some append mode manipulating the
+        # file pointer and with json.dumps?
```

Comment:
comment: 
I couldn't find any document, but it's worth checking to see if it supports the jsonl format.
The jsonl format is append friendly.

Reply:
chrome does not support jsonl format for displaying the trace :(